src/pipeline.cpp

```diff
@@ -223,44 +233,24 @@ static PJ_LP pipeline_reverse (PJ_XY xy, PJ *P) {
 
 
 static PJ *destructor (PJ *P, int errlev) {
-    int i;
     if (nullptr==P)
         return nullptr;
 
     if (nullptr==P->opaque)
         return pj_default_destructor (P, errlev);
 
-    /* Deallocate each pipeline step, then pipeline array */
-    if (nullptr!=static_cast<struct pj_opaque*>(P->opaque)->pipeline)
-        for (i = 0;  i < static_cast<struct pj_opaque*>(P->opaque)->steps; i++)
-            proj_destroy (static_cast<struct pj_opaque*>(P->opaque)->pipeline[i+1]);
-    pj_dealloc (static_cast<struct pj_opaque*>(P->opaque)->pipeline);
+    auto opaque = static_cast<struct pj_opaque*>(P->opaque);
 
-    pj_dealloc (static_cast<struct pj_opaque*>(P->opaque)->argv);
-    pj_dealloc (static_cast<struct pj_opaque*>(P->opaque)->current_argv);
+    pj_dealloc (opaque->argv);
+    pj_dealloc (opaque->current_argv);
 
-    for (i=0; i<4; i++)
-        delete static_cast<struct pj_opaque*>(P->opaque)->stack[i];
+    delete opaque;
```

Comment:
Is this safe? I admit that I do not have the full overview of this PR, but shouldn't the `*P` destructor handle this, for cases where additional memory was allocated at construction time?

Reply:
yes, this is safe because I take care just the line below to nullify P->opaque, so that generic code doesn't try to use free() on it. And use of delete is safe (and needed!) since opaque was allocated by new at https://github.com/OSGeo/PROJ/pull/1748/files#diff-ca6b9218666b6a49063bd0c51b946e35R426 . Allocating with new is needed since the Pipeline struct holds C++ objects (std::vector<Step> steps and std::stack<double> stack[4]) thus those for proper construction and destruction of them
Admitdely the current situation of PROJ operator implementation coming from C, and with some C++ bits added there and there, can be a bit challenging :-). But here I found that using C++ facilities made the code much easier to follow despite that subtelty in the destructor function.

Reply:
@rouault thanks for the explanation. A full C++ core certainly would improve the general readability of the code, but that is probably many years away :-)